validators/valset/snapshot/set.go

```diff
@@ -0,0 +1,538 @@
+		beginHeight := currentEpoch * s.epochSize
+		beginHeader, ok := s.blockchain.GetHeaderByNumber(beginHeight)
+
+		if !ok {
+			return fmt.Errorf("header at %d not found", beginHeight)
+		}
+
+		if err := s.addHeaderSnap(beginHeader); err != nil {
+			return err
+		}
+
+		s.store.updateLastBlock(beginHeight)
+
+		meta = s.GetSnapshotMetadata()
+	}
+
+	// Process headers if we missed some blocks in the current epoch
+	if header.Number > meta.LastBlock {
+		s.logger.Info("syncing past snapshots", "from", meta.LastBlock, "to", header.Number)
+
+		for num := meta.LastBlock + 1; num <= header.Number; num++ {
+			if num == 0 {
+				continue
+			}
+
+			header, ok := s.blockchain.GetHeaderByNumber(num)
+			if !ok {
+				return fmt.Errorf("header %d not found", num)
+			}
+
+			if err := s.ProcessHeader(header); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) SourceType() valset.SourceType {
+	return valset.Snapshot
+}
+
+func (s *SnapshotValidatorSet) GetValidators(height uint64) (validators.Validators, error) {
+	var snapshotHeight uint64 = 0
+	if int64(height) > 0 {
+		snapshotHeight = height - 1
+	}
+
+	snapshot := s.getSnapshot(snapshotHeight)
+	if snapshot == nil {
+		return nil, ErrSnapshotNotFound
+	}
+
+	return snapshot.Set, nil
+}
+
+func (s *SnapshotValidatorSet) GetSnapshotMetadata() *SnapshotMetadata {
+	return &SnapshotMetadata{
+		LastBlock: s.store.getLastBlock(),
+	}
+}
+
+func (s *SnapshotValidatorSet) GetSnapshots() []*Snapshot {
+	return s.store.list
+}
+
+func (s *SnapshotValidatorSet) UpdateSet(newValidators validators.Validators, from uint64) error {
+	snapshotHeight := from - 1
+
+	snapshot := s.getSnapshot(snapshotHeight)
+	if snapshot == nil {
+		snapshot = &Snapshot{}
+	}
+
+	newSnapshot := snapshot.Copy()
+
+	header, _ := s.blockchain.GetHeaderByNumber(from - 1)
+	if header == nil {
+		return fmt.Errorf("header at %d not found", from-1)
+	}
+
+	newSnapshot.Number = snapshotHeight
+	newSnapshot.Hash = header.Hash.String()
+	newSnapshot.Set = newValidators
+	newSnapshot.Votes = []*valset.Vote{}
+
+	if !newSnapshot.Equal(snapshot) {
+		s.store.add(newSnapshot)
+	}
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) ModifyHeader(header *types.Header, proposer types.Address) error {
+	snapshot := s.getSnapshot(header.Number)
+	if snapshot == nil {
+		return ErrSnapshotNotFound
+	}
+
+	if candidate := s.getNextCandidate(snapshot, proposer); candidate != nil {
+		header.Miner = types.MarshalRLPTo(candidate.Validator.MarshalRLPWith, nil)
+
+		if candidate.Authorize {
+			header.Nonce = nonceAuthVote
+		} else {
+			header.Nonce = nonceDropVote
+		}
+	}
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) VerifyHeader(header *types.Header) error {
+	if header.Nonce != nonceAuthVote && header.Nonce != nonceDropVote {
+		return ErrInvalidNonce
+	}
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) ProcessHeader(
```

Comment:
Does this method really need to be 300 lines long?

Reply:
Refactored at 5e5464e32b342499efed390c0579c7c8bce11021

Reply:
Much cleaner, thank you 🙏